### DISCUSSION TITLE
Convert casadi function to normal matlab/c function

### DISCUSSION THREAD

**Discussion by zhangjiadi44** [casadi/casadi#3337] [Q&A] [answered]
Is there any way to convert the symbolic expression of Casadi Function in Matlab to a normal matlab function (.m file) or a c function that can be used on MATLAB real-time?

**Comment by jgillis** [comment, marked as answer]
CasADi features mature code generation. If yout Function is `f`, just do `f.generate('f.c')`. See user guide and codegen examples for more info. There was some kind of export to Matlab code at some point too: `f.export_code('matlab','f')` This ia experimental and should be used with caution.

**Reply by npn0010** [reply]
I have a follow up question related to this. Is there a way to convert a `casadi.Function` to a MATLAB `function_handle` (i.e., like an anonymous function) instead of generating a new .m file? My goal is to be able to generate a function and then use it in a MATLAB function like `bvp4c` in the same script.

**Reply by jgillis** [reply]
`@(x) full(f(x))` may work

**Reply by npn0010** [reply]
Works great. Thanks so much.

**Reply by npn0010** [reply]
What about an anonymous function that has multiple outputs?  I get an error saying `One or more output arguments not assigned during call to "casadiMEX".` when trying multiple outputs with `@(x) full(f(x))`.

**Reply by zhangjiadi44** [reply]
Using `f.export_code('matlab','f')` , it can generate matlab function. But it will pass all the inputs to the `nonzeros_gen()`  function first, which is defined in the same .m file as
`function y=nonzeros_gen(x)
  if isa(x,'casadi.SX') || isa(x,'casadi.MX') || isa(x,'casadi.DM')
    y = x{:};
  elseif isa(x,'sdpvar')
    b = getbase(x);
    f = find(sum(b~=0,2));
    y = sdpvar(length(f),1,[],getvariables(x),b(f,:));
  else
    y = nonzeros(x);
  end
end` 

But the input can be a numerical vector with an element equals to zero somewhere. Thus, if it's passed to `nonzeros_gen()` the dimension will be wrong for the following calculation. Is there any way to modify the `f.export_code('matlab','f')` function such that the input will not pass through `nonzeros_gen()`?

**Reply by npn0010** [reply]
Is there any fix to this?

**Comment by charleskhazoom** [comment]
Here is a Matlab function that takes a casadi function, generates `.c` code and makes a mex function from the `.c` code. The mex function can then be ran in Matlab.
```
function mex_name = casadi_fun_to_mex(casadi_fun, dir, opt_flag)
    arguments
        casadi_fun (1,1) casadi.Function
        dir char
        opt_flag = '-O3'
    end

    disp(['Compiling with ',opt_flag])
    
    mex_name = casadi_fun.name;
    full_name = [dir, '/', mex_name, '.c'];
    opts = struct('main', true, ...
        'mex', true);
    casadi_fun.generate(mex_name, opts);
    if ~strcmp(dir,pwd)
        movefile([mex_name,'.c'], dir);
    end
    mex(full_name, '-largeArrayDims', ['COPTIMFLAGS="',opt_flag,'"']);
    
    if ~strcmp(dir,pwd)
        movefile([mex_name,'.mexa64'], dir);
    end
end
```
